### ISSUE TITLE
[BUG] various performance issues

### ISSUE THREAD

**Issue by davidgilbertson** [sktime/sktime#4807]
**Describe the bug**
I'm in the process of porting my existing time series forecasting project over to sktime. I've now got both versions of the project running side by side on the same data, getting the same results (ish). I have multiple instances (a hierarchical index), so I'm training one model (Gradient Boosting Regressor) per instance.

The run times are wildly different in two main areas:

## 1. Splitting train/test

My version: 52ms
Sktime: 3,922ms

There's only 8 instances, and a total of ~13,000 rows.

I haven't looked hard for the problem, but one guess is that this iteration is slow: https://github.com/sktime/sktime/blob/main/sktime/forecasting/model_selection/_split.py#L1389-L1392

I think if you changed `yrl1 = yret[0].to_list()` to `yrl1 = set(yret[0].to_list())` it would be must faster. But I haven't really thought about what this code is doing, there's surely a faster way than looping. Like, if the user wants the last 16 dates of each instance in a test split, then:
```py
test = df.groupby(df.index.names[:-1]).tail(16)  # last 16 of each instance
train = df.drop(tst.index)  # everything else is training
```
(with appropriate checks for monotonicity, etc)

## 2. Fitting/predicting
My version is manually looping over groups of the dataframe, fitting a model for each instance and doing the predictions, stepping over the (16) horizon steps and all that fun stuff. I assume much the same as what sktime is doing.

My version: 1.4s
Sktime: 10.0s

I'm not sure what sktime is doing for an additional 8+ seconds. And this number varies a _lot_. Sometimes it 10s, one time 54s, occasionally 2.5s, for the exact same script. My version is always within about 0.2s of 1.4. I have no idea what's causing this.

For the times reported above my pipeline was using `YfromX`, but I've noticed similarly slow performance using `make_reduction` and `RecursiveTabularRegressionForecaster`. Only when I have `pooling="global"` is sktime faster.

Sorry I don't have time to dive in further or do a proper PR, hopefully this is useful info as is.

**Comment by fkiraly** [comment]
Thanks for reporting!

Some comments about the cases above:

1. `temporal_train_test_split` is quite old and has been patched here and there for hierarchical data. A clean re-write would probably be warranted.
2. re fitting/predicting, can you kindly dump your code? With dummy data preferably. If it is not too much of a hassle.

**Comment by davidgilbertson** [comment]
I tried slimming down the code I had, but found that running it in a new file was much faster, so I can't replicate it taking 10 second to run, which is odd because it definitely took that long over at least 5 runs yesterday. I also found that if I ran the sktime code in an interpreter that had already run "my version" code, it finished faster. Just having `torch` imported makes sktime run half a second faster (and nope, I'm not actually using PyTorch anywhere!), but there was something else going on I couldn't get to the bottom of, and now it's running fast. I'll comment further if I see the issue again.

**Comment by fkiraly** [comment]
Hm, sounds like an environment or imports issue, @davidgilbertson?
Sometimes, the first batch of imports can take long.
